[Bug]: Nessie-GC docker image ClassNotFoundException: software.amazon.awssdk.http.apache.ApacheHttpClient$Builder

### What happened

I have been using nessie and nessie GC for at least 18 months, using postgresql and an S3 like storage (netapp implementation) without any problem and following the upgrade cycles about every 1-2 months.

upgrading from 0.107.5 to 0.108.3, my GC runs which use the ghcr.io/projectnessie/nessie-gc docker image suddenly failed with a ClassNotFoundException: software.amazon.awssdk.http.apache.ApacheHttpClient$Builder

Trying to bissect the releases I skipped to narrow down the exact failing version, I believe it the problem to have appeared since 0.107.6 and and 0.107.9.

Claude 2 cents comment is

> That's exactly the pattern you'd expect from a transitive AWS SDK BOM bump slipping in silently via a renovate/dependabot PR, with no manual review of behavior changes.
> 
> And critically: AWS SDK for Java 2.x made Apache 5 the default HTTP client starting in v2.46.0, released June 1, 2026 — which falls squarely between your working 0.107.5/0.107.6 (May 27) and broken 0.107.9 (published ~2 months before today, so early-to-mid June 2026).

but no idea what extent of hallucination it has.

Not sure what to add.

Best regards,

### How to reproduce it

See additional information for the docker CLI starting flags as well as the description regarding the environment 

### Nessie server type (docker/uber-jar/built from source) and version

docker 0.107.9

### Client type (Ex: UI/Spark/pynessie ...) and version

_No response_

### Additional information

```
2026.07.31 07:02:05.214 docker        run
        --rm
        -i
        --name "nessie-gc-bc"
        --link="postgresql:psql"
        --link="nessie-bc:nessie"
                 -e NESSIE_GC_ICEBERG="s3.access-key-id=XXXXXXXXXXXXXXXXXXXX,s3.secret-access-key=YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY,s3.endpoint=https://dansons.lacarioka.com,s3.path-style-access=true"
        -e NESSIE_GC_PLAIN_OUTPUT=true
        -e AWS_REGION=local
        -e AWS_ENDPOINT_URL=https://dansons.lacarioka.com
        -e JAVA_TOOL_OPTIONS="-Dlogback.statusListenerClass=ch.qos.logback.core.status.NopStatusListener"         -v "/etc/ssl/certs/java/cacerts:/etc/pki/ca-trust/extracted/java/cacerts:ro"
        -v "/backup/nessie-gc-workdir/tmp.sX6aOCsDwk/sharedir:/backup/nessie-gc-workdir/tmp.sX6aOCsDwk/sharedir"         "ghcr.io/projectnessie/nessie-gc:0.108.3"
        sweep

        --jdbc-url="jdbc:postgresql://psql:5432/bc?user=nessie"
        --jdbc-password="UUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUU"
        --defer-deletes --expiry-parallelism=1 --live-set-id=d82ceac5-c270-473e-bfae-848ef8ffb21a
WARNING: Links on the default bridge network are deprecated and will be removed in a future release. Use a custom network instead.
Picked up JAVA_TOOL_OPTIONS: -Dlogback.statusListenerClass=ch.qos.logback.core.status.NopStatusListener
Exception in thread "main" java.lang.NoClassDefFoundError
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:542)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:567)
	at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:653)
	at java.base/java.util.concurrent.ForkJoinPool.invoke(ForkJoinPool.java:2822)
	at org.projectnessie.gc.expire.local.DefaultLocalExpire.expire(DefaultLocalExpire.java:75)
	at org.projectnessie.gc.tool.cli.commands.BaseRepositoryCommand.expire(BaseRepositoryCommand.java:250)
	at org.projectnessie.gc.tool.cli.commands.Sweep.call(Sweep.java:43)
	at org.projectnessie.gc.tool.cli.commands.BaseLiveSetCommand.call(BaseLiveSetCommand.java:42)
	at org.projectnessie.gc.tool.cli.commands.BaseRepositoryCommand.call(BaseRepositoryCommand.java:84)
	at org.projectnessie.gc.tool.cli.commands.BaseCommand.call(BaseCommand.java:26)
	at org.projectnessie.gc.tool.cli.commands.BaseCommand.call(BaseCommand.java:21)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2045)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
	at picocli.CommandLine.execute(CommandLine.java:2174)
	at org.projectnessie.gc.tool.cli.CLI.runMain(CLI.java:104)
	at org.projectnessie.gc.tool.cli.CLI.runMain(CLI.java:75)
	at org.projectnessie.gc.tool.cli.CLI.main(CLI.java:70)
Caused by: java.lang.NoClassDefFoundError: software/amazon/awssdk/http/apache/ApacheHttpClient$Builder
	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3580)
	at java.base/java.lang.Class.getDeclaredMethod(Class.java:2848)
	at org.apache.iceberg.common.DynMethods$Builder.hiddenImpl(DynMethods.java:376)
	at org.apache.iceberg.common.DynMethods$Builder.hiddenImpl(DynMethods.java:335)
	at org.apache.iceberg.common.DynMethods$Builder.hiddenImpl(DynMethods.java:354)
	at org.apache.iceberg.aws.HttpClientProperties.loadHttpClientConfigurations(HttpClientProperties.java:270)
	at org.apache.iceberg.aws.HttpClientProperties.applyHttpClientConfigurations(HttpClientProperties.java:251)
	at software.amazon.awssdk.utils.builder.SdkBuilder.applyMutation(SdkBuilder.java:61)
	at org.apache.iceberg.aws.AwsClientFactories$DefaultAwsClientFactory.s3(AwsClientFactories.java:112)
	at org.apache.iceberg.aws.s3.PrefixedS3Client.s3(PrefixedS3Client.java:85)
	at org.apache.iceberg.aws.s3.S3InputFile.fromLocation(S3InputFile.java:41)
	at org.apache.iceberg.aws.s3.S3InputFile.fromLocation(S3InputFile.java:35)
	at org.apache.iceberg.aws.s3.S3FileIO.newInputFile(S3FileIO.java:152)
	at org.apache.iceberg.io.ResolvingFileIO.newInputFile(ResolvingFileIO.java:90)
	at org.apache.iceberg.TableMetadataParser.read(TableMetadataParser.java:294)
	at org.projectnessie.gc.iceberg.IcebergContentToFiles.extractTableFiles(IcebergContentToFiles.java:137)
	at org.projectnessie.gc.iceberg.IcebergContentToFiles.extractFiles(IcebergContentToFiles.java:105)
	at org.projectnessie.gc.expire.PerContentDeleteExpired.lambda$identifyLiveFiles$2(PerContentDeleteExpired.java:139)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:273)
	at org.projectnessie.gc.contents.jdbc.JdbcHelper$ResultSetSplit.tryAdvance(JdbcHelper.java:151)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at org.projectnessie.gc.expire.PerContentDeleteExpired.identifyLiveFiles(PerContentDeleteExpired.java:145)
	at org.projectnessie.gc.expire.PerContentDeleteExpired.expire(PerContentDeleteExpired.java:73)
	at org.projectnessie.gc.expire.local.DefaultLocalExpire.expireSingleContent(DefaultLocalExpire.java:109)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:960)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:934)
	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.helpComplete(ForkJoinPool.java:1395)
	at java.base/java.util.concurrent.ForkJoinPool.helpComplete(ForkJoinPool.java:2104)
	at java.base/java.util.concurrent.ForkJoinTask.awaitDone(ForkJoinTask.java:420)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:668)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:927)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:657)
	at org.projectnessie.gc.expire.local.DefaultLocalExpire.expireInForkJoinPool(DefaultLocalExpire.java:96)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1456)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.lang.ClassNotFoundException: software.amazon.awssdk.http.apache.ApacheHttpClient$Builder
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 54 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,6 +20,9 @@ as necessary. Empty sections will not end in the release notes.
 
 ### Fixes
 
+- GC: Nessie GC was missing an explicit dependency to Apache HTTP client v4, which is required for Iceberg.
+  The AWSSDK update to 2.46.0 in 0.107.9 changed the default to Apache client v5, so v4 was missing.
+
 ### Commits
 
 ## [0.108.3] Release (2026-07-23)

--- a/gc/gc-iceberg-files/build.gradle.kts
+++ b/gc/gc-iceberg-files/build.gradle.kts
@@ -76,7 +76,11 @@ dependencies {
   testFixturesApi(platform(libs.awssdk.bom))
   testFixturesApi("software.amazon.awssdk:s3")
   testFixturesRuntimeOnly("software.amazon.awssdk:sts")
-  testFixturesRuntimeOnly(libs.hadoop.aws)
+  testFixturesRuntimeOnly("software.amazon.awssdk:kms")
+  testFixturesRuntimeOnly(libs.hadoop.aws) {
+    exclude("software.amazon.awssdk", "bundle")
+  }
+  testFixturesRuntimeOnly("software.amazon.awssdk:apache-client")
 
   testFixturesRuntimeOnly(platform(libs.google.cloud.storage.bom))
   testFixturesRuntimeOnly(platform(libs.google.cloud.libraries.bom))

--- a/gc/gc-iceberg-files/src/test/java/org/projectnessie/gc/iceberg/files/TestIcebergS3Files.java
+++ b/gc/gc-iceberg-files/src/test/java/org/projectnessie/gc/iceberg/files/TestIcebergS3Files.java
@@ -37,7 +37,6 @@ public class TestIcebergS3Files extends AbstractFiles {
     props.put("s3.endpoint", server.getS3BaseUri().toString());
     // must enforce path-style access because S3Resource has the bucket name in its path
     props.put("s3.path-style-access", "true");
-    props.put("http-client.type", "urlconnection");
 
     return props;
   }

--- a/gc/gc-iceberg-inttest/build.gradle.kts
+++ b/gc/gc-iceberg-inttest/build.gradle.kts
@@ -100,16 +100,18 @@ dependencies {
   )
 
   intTestRuntimeOnly(libs.hadoop.client)
-  intTestRuntimeOnly(libs.hadoop.aws)
-  intTestRuntimeOnly("software.amazon.awssdk:sts")
+  intTestRuntimeOnly(libs.hadoop.aws) {
+    exclude("software.amazon.awssdk", "bundle")
+  }
 
   intTestImplementation(platform(libs.awssdk.bom))
   intTestImplementation("software.amazon.awssdk:s3")
-  runtimeOnly("software.amazon.awssdk:url-connection-client")
-  // TODO those are needed, because Spark serializes some configuration stuff (Spark broadcast)
+  intTestRuntimeOnly("software.amazon.awssdk:s3-transfer-manager")
   intTestRuntimeOnly("software.amazon.awssdk:dynamodb")
   intTestRuntimeOnly("software.amazon.awssdk:glue")
   intTestRuntimeOnly("software.amazon.awssdk:kms")
+  intTestRuntimeOnly("software.amazon.awssdk:sts")
+  intTestRuntimeOnly("software.amazon.awssdk:apache-client")
 
   intTestImplementation(platform(libs.google.cloud.storage.bom))
   intTestRuntimeOnly(platform(libs.google.cloud.libraries.bom))

--- a/gc/gc-tool-inttest/build.gradle.kts
+++ b/gc/gc-tool-inttest/build.gradle.kts
@@ -93,7 +93,9 @@ dependencies {
   )
 
   intTestRuntimeOnly(libs.hadoop.client)
-  intTestRuntimeOnly(libs.hadoop.aws)
+  intTestRuntimeOnly(libs.hadoop.aws) {
+    exclude("software.amazon.awssdk", "bundle")
+  }
   intTestRuntimeOnly("software.amazon.awssdk:sts")
 
   intTestRuntimeOnly(platform(libs.awssdk.bom))

--- a/gc/gc-tool/build.gradle.kts
+++ b/gc/gc-tool/build.gradle.kts
@@ -71,6 +71,8 @@ dependencies {
 
   implementation(platform(libs.awssdk.bom))
   runtimeOnly("software.amazon.awssdk:s3")
+  runtimeOnly("software.amazon.awssdk:apache-client")
+  runtimeOnly("software.amazon.awssdk:apache5-client")
   runtimeOnly("software.amazon.awssdk:url-connection-client")
   runtimeOnly("software.amazon.awssdk:sts")
   runtimeOnly("software.amazon.awssdk:kms")

--- a/integrations/spark-extensions/build.gradle.kts
+++ b/integrations/spark-extensions/build.gradle.kts
@@ -51,11 +51,19 @@ dependencies {
   testFixturesImplementation("org.apache.iceberg:iceberg-spark-extensions-${sparkScala.sparkMajorVersion}_${sparkScala.scalaMajorVersion}:$versionIceberg")
   testFixturesImplementation("org.apache.iceberg:iceberg-hive-metastore:$versionIceberg")
   testFixturesImplementation("org.apache.iceberg:iceberg-aws:$versionIceberg")
-  testFixturesImplementation("org.apache.iceberg:iceberg-aws-bundle:$versionIceberg")
   intTestRuntimeOnly(libs.hadoop.client) {
     exclude("org.slf4j", "slf4j-reload4j")
   }
-  intTestRuntimeOnly(libs.hadoop.aws)
+  intTestRuntimeOnly(libs.hadoop.aws) {
+    exclude("software.amazon.awssdk", "bundle")
+  }
+  intTestRuntimeOnly(platform(libs.awssdk.bom))
+  intTestRuntimeOnly("software.amazon.awssdk:apache-client")
+  intTestRuntimeOnly("software.amazon.awssdk:s3-transfer-manager")
+  intTestRuntimeOnly("software.amazon.awssdk:dynamodb")
+  intTestRuntimeOnly("software.amazon.awssdk:glue")
+  intTestRuntimeOnly("software.amazon.awssdk:kms")
+  intTestRuntimeOnly("software.amazon.awssdk:sts")
 
   testFixturesRuntimeOnly(libs.logback.classic)
   testFixturesImplementation(libs.slf4j.log4j.over.slf4j) {

--- a/testing/floci-s3-container/src/intTest/java/org/projectnessie/testing/floci/s3/ITFlociS3Extension.java
+++ b/testing/floci-s3-container/src/intTest/java/org/projectnessie/testing/floci/s3/ITFlociS3Extension.java
@@ -50,7 +50,7 @@ public class ITFlociS3Extension {
         .containsEntry("s3.access-key-id", flociS3.accessKey())
         .containsEntry("s3.secret-access-key", flociS3.secretKey())
         .containsEntry("s3.endpoint", flociS3.s3endpoint())
-        .containsKey("http-client.type");
+        .doesNotContainKey("http-client.type");
 
     soft.assertThat(flociS3.hadoopConfig())
         .isNotNull()

--- a/testing/floci-s3-container/src/main/java/org/projectnessie/testing/floci/s3/FlociS3Container.java
+++ b/testing/floci-s3-container/src/main/java/org/projectnessie/testing/floci/s3/FlociS3Container.java
@@ -136,7 +136,6 @@ public final class FlociS3Container extends FlociContainer implements FlociS3Acc
     props.put("s3.secret-access-key", secretKey());
     props.put("s3.endpoint", s3endpoint());
     props.put("s3.path-style-access", "true");
-    props.put("http-client.type", "urlconnection");
     return props;
   }
 

